File: Tests/VIPSImageSavingTests.m
//
//  VIPSImageSavingTests.m
//  VIPSKitTests
//
//  Tests for VIPSImage+Saving: file saving and data export methods
//

#import "VIPSImageTestCase.h"

@interface VIPSImageSavingTests : VIPSImageTestCase
@end

@implementation VIPSImageSavingTests

#pragma mark - JPEG Export Tests

- (void)testExportToJPEG {
    VIPSImage *image = [self createTestImageWithWidth:100 height:100];

    NSError *error = nil;
    NSData *jpegData = [image dataWithFormat:VIPSImageFormatJPEG quality:85 error:&error];

    XCTAssertNotNil(jpegData, @"Should export to JPEG");
    XCTAssertNil(error, @"Should not have error: %@", error);
    XCTAssertGreaterThan(jpegData.length, 0, @"JPEG data should not be empty");

    // Check JPEG magic bytes
    const unsigned char *bytes = (const unsigned char *)jpegData.bytes;
    XCTAssertEqual(bytes[0], 0xFF, @"Should have JPEG magic byte 1");
    XCTAssertEqual(bytes[1], 0xD8, @"Should have JPEG magic byte 2");
}

- (void)testExportToJPEGQualityRange {
    VIPSImage *image = [self createTestImageWithWidth:100 height:100];
    NSError *error = nil;

    // Test low quality
    NSData *lowQuality = [image dataWithFormat:VIPSImageFormatJPEG quality:10 error:&error];
    XCTAssertNotNil(lowQuality, @"Should export low quality JPEG");

    // Test high quality
    NSData *highQuality = [image dataWithFormat:VIPSImageFormatJPEG quality:100 error:&error];
    XCTAssertNotNil(highQuality, @"Should export high quality JPEG");

    // High quality should generally be larger
    XCTAssertGreaterThan(highQuality.length, lowQuality.length, @"High quality should be larger");
}

#pragma mark - PNG Export Tests

- (void)testExportToPNG {
    VIPSImage *image = [self createTestImageWithWidth:100 height:100];

    NSError *error = nil;
    NSData *pngData = [image dataWithFormat:VIPSImageFormatPNG quality:0 error:&error];

    XCTAssertNotNil(pngData, @"Should export to PNG");
    XCTAssertNil(error, @"Should not have error: %@", error);

    // Check PNG magic bytes
    const unsigned char *bytes = (const unsigned char *)pngData.bytes;
    XCTAssertEqual(bytes[0], 0x89, @"Should have PNG magic byte 1");
    XCTAssertEqual(bytes[1], 'P', @"Should have PNG magic byte 2");
    XCTAssertEqual(bytes[2], 'N', @"Should have PNG magic byte 3");
    XCTAssertEqual(bytes[3], 'G', @"Should have PNG magic byte 4");
}

- (void)testExportToPNGWithAlpha {
    VIPSImage *image = [self createTestImageWithWidth:100 height:100 bands:4];

    NSError *error = nil;
    NSData *pngData = [image dataWithFormat:VIPSImageFormatPNG quality:0 error:&error];

    XCTAssertNotNil(pngData, @"Should export RGBA image to PNG");
    XCTAssertNil(error, @"Should not have error: %@", error);
}

#pragma mark - WebP Export Tests

- (void)testExportToWebP {
    VIPSImage *image = [self createTestImageWithWidth:100 height:100];

    NSError *error = nil;
    NSData *webpData = [image dataWithFormat:VIPSImageFormatWebP quality:80 error:&error];

    XCTAssertNotNil(webpData, @"Should export to WebP");
    XCTAssertNil(error, @"Should not have error: %@", error);

    // Check WebP magic bytes (RIFF....WEBP)
    const unsigned char *bytes = (const unsigned char *)webpData.bytes;
    XCTAssertEqual(bytes[0], 'R', @"Should have RIFF header");
    XCTAssertEqual(bytes[1], 'I', @"Should have RIFF header");
    XCTAssertEqual(bytes[2], 'F', @"Should have RIFF header");
    XCTAssertEqual(bytes[3], 'F', @"Should have RIFF header");
}

#pragma mark - File Saving Tests

- (void)testWriteToFile {
    VIPSImage *image = [self createTestImageWithWidth:100 height:100];

    NSString *tempPath = [NSTemporaryDirectory() stringByAppendingPathComponent:@"test_output.jpg"];
    NSError *error = nil;

    BOOL success = [image writeToFile:tempPath error:&error];

    XCTAssertTrue(success, @"Should write to file: %@", error);
    XCTAssertNil(error, @"Should not have error");

    // Verify file exists
    BOOL exists = [[NSFileManager defaultManager] fileExistsAtPath:tempPath];
    XCTAssertTrue(exists, @"File should exist");

    // Cleanup
    [[NSFileManager defaultManager] removeItemAtPath:tempPath error:nil];
}

- (void)testWriteToFileWithFormat {
    VIPSImage *image = [self createTestImageWithWidth:100 height:100];

    NSString *tempPath = [NSTemporaryDirectory() stringByAppendingPathComponent:@"test_output.png"];
    NSError *error = nil;

    BOOL success = [image writeToFile:tempPath format:VIPSImageFormatPNG quality:0 error:&error];

    XCTAssertTrue(success, @"Should write PNG to file: %@", error);
    XCTAssertNil(error, @"Should not have error");

    // Verify file exists and has PNG magic bytes
    NSData *data = [NSData dataWithContentsOfFile:tempPath];
    XCTAssertNotNil(data, @"Should read file");

    const unsigned char *bytes = (const unsigned char *)data.bytes;
    XCTAssertEqual(bytes[0], 0x89, @"Should have PNG magic byte");

    // Cleanup
    [[NSFileManager defaultManager] removeItemAtPath:tempPath error:nil];
}

#pragma mark - Roundtrip Tests

- (void)testJPEGRoundtrip {
    VIPSImage *original = [self createTestImageWithWidth:100 height:100];

    // Export to JPEG
    NSError *error = nil;
    NSData *jpegData = [original dataWithFormat:VIPSImageFormatJPEG quality:95 error:&error];
    XCTAssertNotNil(jpegData, @"Should export to JPEG");

    // Load back
    VIPSImage *loaded = [VIPSImage imageWithData:jpegData error:&error];
    XCTAssertNotNil(loaded, @"Should load JPEG data");
    XCTAssertEqual(loaded.width, original.width, @"Width should match");
    XCTAssertEqual(loaded.height, original.height, @"Height should match");
}

- (void)testPNGRoundtrip {
    VIPSImage *original = [self createTestImageWithWidth:100 height:100 bands:4];

    // Export to PNG
    NSError *error = nil;
    NSData *pngData = [original dataWithFormat:VIPSImageFormatPNG quality:0 error:&error];
    XCTAssertNotNil(pngData, @"Should export to PNG");

    // Load back
    VIPSImage *loaded = [VIPSImage imageWithData:pngData error:&error];
    XCTAssertNotNil(loaded, @"Should load PNG data");
    XCTAssertEqual(loaded.width, original.width, @"Width should match");
    XCTAssertEqual(loaded.height, original.height, @"Height should match");
    XCTAssertEqual(loaded.bands, original.bands, @"Bands should match");
}

@end

File: Tests/VIPSImageSavingTests.swift
import XCTest
@testable import VIPSKit

final class VIPSImageSavingTests: VIPSImageTestCase {

    func testSaveJPEG() throws {
        let image = createTestImage(width: 100, height: 100)
        let data = try image.data(format: .jpeg, quality: 85)
        XCTAssertGreaterThan(data.count, 0)
        // Check JPEG magic bytes
        XCTAssertEqual(data[0], 0xFF)
        XCTAssertEqual(data[1], 0xD8)
    }

    func testSavePNG() throws {
        let image = createTestImage(width: 100, height: 100)
        let data = try image.data(format: .png)
        XCTAssertGreaterThan(data.count, 0)
        // Check PNG magic
        XCTAssertEqual(data[0], 0x89)
        XCTAssertEqual(data[1], 0x50) // 'P'
    }

    func testSaveWebP() throws {
        let image = createTestImage(width: 100, height: 100)
        let data = try image.data(format: .webP, quality: 80)
        XCTAssertGreaterThan(data.count, 0)
    }

    func testSaveJPEGQuality() throws {
        let image = createTestImage(width: 100, height: 100)
        let lowQ = try image.data(format: .jpeg, quality: 10)
        let highQ = try image.data(format: .jpeg, quality: 95)
        XCTAssertLessThan(lowQ.count, highQ.count)
    }

    func testWriteToFile() throws {
        let image = createTestImage(width: 100, height: 100)
        let path = NSTemporaryDirectory() + "test_save.jpg"
        try image.write(toFile: path)
        XCTAssertTrue(FileManager.default.fileExists(atPath: path))
        try? FileManager.default.removeItem(atPath: path)
    }

    func testWriteToFileWithFormat() throws {
        let image = createTestImage(width: 100, height: 100)
        let path = NSTemporaryDirectory() + "test_save_explicit.png"
        try image.write(toFile: path, format: .png)
        XCTAssertTrue(FileManager.default.fileExists(atPath: path))
        try? FileManager.default.removeItem(atPath: path)
    }

    func testRoundtripJPEG() throws {
        let source = createTestImage(width: 100, height: 100)
        let data = try source.data(format: .jpeg, quality: 100)
        let loaded = try VIPSImage(data: data)
        XCTAssertEqual(loaded.width, source.width)
        XCTAssertEqual(loaded.height, source.height)
    }

    func testRoundtripPNG() throws {
        let source = createTestImage(width: 100, height: 100)
        let data = try source.data(format: .png)
        let loaded = try VIPSImage(data: data)
        XCTAssertEqual(loaded.width, source.width)
        XCTAssertEqual(loaded.height, source.height)
    }

    func testRoundtripWebP() throws {
        let source = createTestImage(width: 100, height: 100)
        let data = try source.data(format: .webP, quality: 90)
        let loaded = try VIPSImage(data: data)
        XCTAssertEqual(loaded.width, source.width)
        XCTAssertEqual(loaded.height, source.height)
        XCTAssertEqual(loaded.sourceFormat, .webP)
    }

    // MARK: - Round-trip from Real Images

    func testRoundtripRealPNGToJPEG() throws {
        guard let path = pathForTestResource("test-rgb.png") else {
            XCTFail("Test resource not found")
            return
        }
        let source = try VIPSImage(contentsOfFile: path)
        let jpegData = try source.data(format: .jpeg, quality: 90)
        let loaded = try VIPSImage(data: jpegData)
        XCTAssertEqual(loaded.width, source.width)
        XCTAssertEqual(loaded.height, source.height)
        XCTAssertEqual(loaded.sourceFormat, .jpeg)
    }

    func testRoundtripRealPNGToWebP() throws {
        guard let path = pathForTestResource("test-rgb.png") else {
            XCTFail("Test resource not found")
            return
        }
        let source = try VIPSImage(contentsOfFile: path)
        let webpData = try source.data(format: .webP, quality: 85)
        let loaded = try VIPSImage(data: webpData)
        XCTAssertEqual(loaded.width, source.width)
        XCTAssertEqual(loaded.height, source.height)
        XCTAssertEqual(loaded.sourceFormat, .webP)
    }

    func testRoundtripWebPToJPEG() throws {
        guard let path = pathForTestResource("test.webp") else {
            XCTFail("Test resource not found")
            return
        }
        let source = try VIPSImage(contentsOfFile: path)
        let jpegData = try source.data(format: .jpeg, quality: 90)
        let loaded = try VIPSImage(data: jpegData)
        XCTAssertEqual(loaded.width, source.width)
        XCTAssertEqual(loaded.height, source.height)
        XCTAssertEqual(loaded.sourceFormat, .jpeg)
    }

    func testRoundtripGIFToPNG() throws {
        guard let path = pathForTestResource("test.gif") else {
            XCTFail("Test resource not found")
            return
        }
        let source = try VIPSImage(contentsOfFile: path)
        let pngData = try source.data(format: .png)
        let loaded = try VIPSImage(data: pngData)
        XCTAssertEqual(loaded.width, source.width)
        XCTAssertEqual(loaded.height, source.height)
        XCTAssertEqual(loaded.sourceFormat, .png)
    }

    // MARK: - Alpha Preservation

    func testSavePNGPreservesAlpha() throws {
        guard let path = pathForTestResource("test-rgba.png") else {
            XCTFail("Test resource not found")
            return
        }
        let source = try VIPSImage(contentsOfFile: path)
        XCTAssertTrue(source.hasAlpha)

        let data = try source.data(format: .png)
        let loaded = try VIPSImage(data: data)
        XCTAssertTrue(loaded.hasAlpha)
        XCTAssertEqual(loaded.bands, 4)
    }

    func testSaveWebPPreservesAlpha() throws {
        guard let path = pathForTestResource("test-rgba.png") else {
            XCTFail("Test resource not found")
            return
        }
        let source = try VIPSImage(contentsOfFile: path)
        XCTAssertTrue(source.hasAlpha)

        let data = try source.data(format: .webP, quality: 90)
        let loaded = try VIPSImage(data: data)
        XCTAssertTrue(loaded.hasAlpha)
        XCTAssertEqual(loaded.bands, 4)
    }

    // MARK: - Grayscale Handling

    func testSaveGrayscaleToJPEG() throws {
        guard let path = pathForTestResource("grayscale.jpg") else {
            XCTFail("Test resource not found")
            return
        }
        let source = try VIPSImage(contentsOfFile: path)
        XCTAssertEqual(source.bands, 1)

        let data = try source.data(format: .jpeg, quality: 85)
        let loaded = try VIPSImage(data: data)
        XCTAssertEqual(loaded.width, source.width)
        XCTAssertEqual(loaded.height, source.height)
    }

    func testSaveGrayscaleToPNG() throws {
        guard let path = pathForTestResource("grayscale.jpg") else {
            XCTFail("Test resource not found")
            return
        }
        let source = try VIPSImage(contentsOfFile: path)
        XCTAssertEqual(source.bands, 1)

        let data = try source.data(format: .png)
        let loaded = try VIPSImage(data: data)
        XCTAssertEqual(loaded.width, source.width)
        XCTAssertEqual(loaded.height, source.height)
    }

    // MARK: - Tiny Image Handling

    func testSaveTinyImage() throws {
        guard let path = pathForTestResource("tiny.png") else {
            XCTFail("Test resource not found")
            return
        }
        let source = try VIPSImage(contentsOfFile: path)
        XCTAssertEqual(source.width, 8)
        XCTAssertEqual(source.height, 8)

        // Test saving to various formats
        let jpegData = try source.data(format: .jpeg, quality: 100)
        let pngData = try source.data(format: .png)
        let webpData = try source.data(format: .webP, quality: 100)

        XCTAssertGreaterThan(jpegData.count, 0)
        XCTAssertGreaterThan(pngData.count, 0)
        XCTAssertGreaterThan(webpData.count, 0)
    }

    // MARK: - Async

    func testAsyncWriteToFile() async throws {
        let image = createTestImage(width: 100, height: 100)
        let path = NSTemporaryDirectory() + "test_async_save.jpg"
        try await image.write(toFile: path)
        XCTAssertTrue(FileManager.default.fileExists(atPath: path))
        try? FileManager.default.removeItem(atPath: path)
    }

    func testAsyncWriteToFileWithFormat() async throws {
        let image = createTestImage(width: 100, height: 100)
        let path = NSTemporaryDirectory() + "test_async_save_explicit.png"
        try await image.write(toFile: path, format: .png)
        XCTAssertTrue(FileManager.default.fileExists(atPath: path))
        try? FileManager.default.removeItem(atPath: path)
    }

    func testAsyncEncoded() async throws {
        let image = createTestImage(width: 100, height: 100)
        let data = try await image.encoded(format: .jpeg, quality: 85)
        XCTAssertGreaterThan(data.count, 0)
        // Check JPEG magic bytes
        XCTAssertEqual(data[0], 0xFF)
        XCTAssertEqual(data[1], 0xD8)
    }
}
